src/test/shell/bazel/bazel_test_test.sh

```diff
@@ -714,36 +714,107 @@ sh_test(
-  outputs_zip=bazel-testlogs/dir/test/test.outputs/outputs.zip
   [ -s $outputs_zip ] || fail "$outputs_zip was not present after test"
 
+  # Check that the original undeclared outputs no longer exist.
+  [ -e $output_text ] && fail "$output_text was present after test"
+  [ -e $output_text ] && fail "$output_text was present after test"
+
+
   # Check the contents of the zip file.
   unzip -q "$outputs_zip" -d unzipped_outputs || fail "failed to unzip $outputs_zip"
   cat > expected_text <<EOF
 some text
 EOF
-diff "unzipped_outputs/text.txt" expected_text > d || fail "unzipped_outputs/text.txt differs from expected:$N$(cat d)$N"
+  diff "unzipped_outputs/text.txt" expected_text > d || fail "unzipped_outputs/text.txt differs from expected:$N$(cat d)$N"
   cat > expected_html <<EOF
 <!DOCTYPE html>
 EOF
-diff expected_html "unzipped_outputs/fake.html" > d || fail "unzipped_outputs/fake.html differs from expected:$N$(cat d)$N"
+  diff expected_html "unzipped_outputs/fake.html" > d || fail "unzipped_outputs/fake.html differs from expected:$N$(cat d)$N"
+}
+
+function test_undeclared_outputs_are_not_zipped() {
+  setup_undeclared_outputs_test
+
+  local -r outputs_dir=bazel-testlogs/dir/test/test.outputs
+  local -r outputs_zip=$outputs_dir/outputs.zip  
+  local -r output_text=$outputs_dir/text.txt
+  local -r output_html=$outputs_dir/fake.html
+
+  bazel test -s --nozip_undeclared_test_outputs //dir:test &> $TEST_log || fail "expected success"
+
+  # Newlines are useful around diffs. This helps us get them in bash strings.
+  N=$'\n'
+
+  # Check that the undeclared outputs zip file does not exist.
+  [ -e $outputs_zip ] && fail "$outputs_zip was present after test"
+
+  # Check that the undeclared outputs exist.
+  [ -e $output_text ] || fail "$output_text was not present after test"
+  [ -e $output_text ] || fail "$output_text was not present after test"
+
+  # Check the contents of the undeclared outputs.
+  cat > expected_text <<EOF
+some text
+EOF
+  diff "$outputs_dir/text.txt" expected_text > d || fail "$outputs_dir/text.txt differs from expected:$N$(cat d)$N"
+  cat > expected_html <<EOF
+<!DOCTYPE html>
+EOF
+  diff expected_html "$outputs_dir/fake.html" > d || fail "$outputs_dir/fake.html differs from expected:$N$(cat d)$N"
+}
+
+function test_undeclared_outputs_zipped_then_unzipped() {
+  setup_undeclared_outputs_test
+
+  local -r outputs_dir=bazel-testlogs/dir/test/test.outputs
+  local -r outputs_zip=$outputs_dir/outputs.zip  
+  local -r output_text=$outputs_dir/text.txt
+  local -r output_html=$outputs_dir/fake.html
+
+  bazel test -s //dir:test &> $TEST_log || fail "expected success"
+
+  [ -s $outputs_zip ] || fail "$outputs_zip was not present after test"
+
+  bazel test -s --nozip_undeclared_test_outputs //dir:test &> $TEST_log || fail "expected success"
+
+  [ -s $output_text ] || fail "$output_text was not present after test"
+  [ -s $output_html ] || fail "$output_html was not present after test"
+}
```

Comment:
Will the zip file be deleted after the build? Also considering the "unzipped and zipped" case, will the text and html files be deleted after second build?

Reply:
Let's test all combinations, then.